pkg/fqdn/doc.go

```diff
@@ -12,20 +12,238 @@
+// solution is multi-layered for historical reasons:
+// - Keep a per-Endpoint cache that can be stored to disk and restored on
+//   startup. These caches apply TTL expiration and limit the IP count per domain.
+// - Keep a global cache to combine all this DNS information and send it to the
+//   policy system. This cache applies TTL but not per-domain limits.
+//   This causes a DNS lookup in one endpoint to leak to another!
+// - Track live connections allowed by DNS policy and delay expiring that data
+//   while the connection is open. If the policy itself is removed, however, the
+//   connection is interrupted.
+//
+// The same DNSCache type is used in all cases. DNSCache instances remain
+// consistent if the update order is different and merging multiple caches
+// should be equivalent to applying the constituent updates individually. As a
+// result, DNS data is all inserted into a single global cache from which the
+// policy layer receives information. This is historic and per-Endpoint
+// handling can be added. The data is internally tracked per IP because
+// overlapping DNS responses may have different TTLs for IPs that appear in
+// both.
+// Notes:
+// - The default configurable minimum TTL in the caches is 1 hour. This is
+//   mostly for identity stability, as short TTLs would cause more identity
+//   churn. This is mostly history as CIDR identities now have a near-0
+//   allocation overhead.
+// - DNSCache deletes only currently occur when the cilium API clears the cache
+//   or when the garbage collector evicts entries.
+// - The combination of caches: per-Endpoint and global must manage disparate
+//   behaviours of pods. The worst case scenario is one where one pod makes many
+//   requests to a target with changing IPs (like S3) but another makes few
+//   requests that are long-lived. We need to ensure "fairness" where one does
+//   not starve the other. The limits in the per-Endpoint caches allow this, and
+//   the global cache acts as a collector across different Endpoints (without
+//   restrictions).
+//
+// Expiration of DNS data is handled by the dns-garbage-collector-job controller.
+// Historically, the only expiration was TTL based and the per-Endpoint and
+// global caches would expire data at the same time without added logic.
+// This is not true when we apply per-host IP limits in the cache. These
+// default to 50 IPs for a given domain, per Endpoint. To account for these
+// evictions the controller handles TTL and IP limit evictions. This ensures
+// that the global cache is consistent with the per-Endpoint caches. The result
+// is that the actual expiration is imprecise (TTL especially). The caches mark
+// to-evict data internally and only do so on GC method calls from the
+// controller.
+// When DNS data is evicted from any per-Endpoint cache, for any reason, each
+// IP is retained as a "zombie". These are tracked per IP for that endpoint
```

Comment:
Not sure if we can work a little more clarity into here, something like:

`These "zombies" represent IPs that were previously associated with a resolved DNS name, but the DNS name is no longer known (for example because of TTL expiry). However there may still be an active connection associated with the zombie IP.`

(then the following part describes alive/dead zombies)